File: tools/dotnet/corelib/Region.cs
// System.Drawing.Region (фаза N9): область как программа операций над путями в
// обратной польской записи — фигура кладётся на стек, операция снимает две.
// Так выражается и `a.Union(b)`, где `b` сама собрана из нескольких фигур.
// Маску точек по программе строит растеризатор (`raster::region`), здесь —
// только запись операций и узнавание частого случая «это прямоугольник».

using System.Collections.Generic;
using System.Drawing.Drawing2D;

namespace System.Drawing
{
    public sealed class Region : IDisposable
    {
        private const int KindInfinite = 0;
        private const int KindEmpty = 1;
        private const int KindPath = 2;
        private const int KindCombine = 3;

        private sealed class Step
        {
            internal int Kind;
            internal int Combine;
            internal PointF[] Points;
            internal byte[] Types;
            internal int FillMode;
        }

        private List<Step> steps = new List<Step>();

        public Region()
        {
            MakeInfinite();
        }

        public Region(RectangleF rect)
        {
            steps.Add(RectangleStep(rect));
        }

        public Region(Rectangle rect)
            : this((RectangleF)rect)
        {
        }

        public Region(GraphicsPath path)
        {
            if (path == null)
            {
                throw new ArgumentNullException("path");
            }
            steps.Add(PathStep(path));
        }

        public void Dispose()
        {
        }

        public Region Clone()
        {
            var copy = new Region();
            copy.steps = new List<Step>();
            for (int i = 0; i < steps.Count; i++)
            {
                Step step = steps[i];
                copy.steps.Add(new Step { Kind = step.Kind, Combine = step.Combine, Points = CopyPoints(step.Points), Types = step.Types, FillMode = step.FillMode });
            }
            return copy;
        }

        private static PointF[] CopyPoints(PointF[] points)
        {
            if (points == null)
            {
                return null;
            }
            var copy = new PointF[points.Length];
            Array.CopyItems(points, copy, points.Length);
            return copy;
        }

        private static Step RectangleStep(RectangleF rect)
        {
            // Пустой прямоугольник — пустая область, а не вырожденный путь.
            if (rect.Width <= 0 || rect.Height <= 0)
            {
                return new Step { Kind = KindEmpty };
            }
            return new Step
            {
                Kind = KindPath,
                Points = new PointF[] { new PointF(rect.X, rect.Y), new PointF(rect.X + rect.Width, rect.Y), new PointF(rect.X + rect.Width, rect.Y + rect.Height), new PointF(rect.X, rect.Y + rect.Height) },
                Types = new byte[] { 0, 1, 1, 0x81 },
            };
        }

        private static Step PathStep(GraphicsPath path) =>
            new Step { Kind = KindPath, Points = path.PathPoints, Types = path.PathTypes, FillMode = (int)path.FillMode };

        public void MakeInfinite()
        {
            steps.Clear();
            steps.Add(new Step { Kind = KindInfinite });
        }

        public void MakeEmpty()
        {
            steps.Clear();
            steps.Add(new Step { Kind = KindEmpty });
        }

        private void Combine(List<Step> operand, CombineMode mode)
        {
            if (mode == CombineMode.Replace)
            {
                steps = operand;
                return;
            }
            steps.AddRange(operand);
            steps.Add(new Step { Kind = KindCombine, Combine = (int)mode });
        }

        private static List<Step> One(Step step)
        {
            var list = new List<Step>();
            list.Add(step);
            return list;
        }

        internal void Combine(RectangleF rect, CombineMode mode) => Combine(One(RectangleStep(rect)), mode);

        internal void Combine(GraphicsPath path, CombineMode mode)
        {
            if (path == null)
            {
                throw new ArgumentNullException("path");
            }
            Combine(One(PathStep(path)), mode);
        }

        internal void Combine(Region region, CombineMode mode)
        {
            if (region == null)
            {
                throw new ArgumentNullException("region");
            }
            Combine(region.Clone().steps, mode);
        }

        public void Intersect(Rectangle rect) => Combine(rect, CombineMode.Intersect);

        public void Intersect(RectangleF rect) => Combine(rect, CombineMode.Intersect);

        public void Intersect(GraphicsPath path) => Combine(path, CombineMode.Intersect);

        public void Intersect(Region region) => Combine(region, CombineMode.Intersect);

        public void Union(Rectangle rect) => Combine(rect, CombineMode.Union);

        public void Union(RectangleF rect) => Combine(rect, CombineMode.Union);

        public void Union(GraphicsPath path) => Combine(path, CombineMode.Union);

        public void Union(Region region) => Combine(region, CombineMode.Union);

        public void Xor(Rectangle rect) => Combine(rect, CombineMode.Xor);

        public void Xor(RectangleF rect) => Combine(rect, CombineMode.Xor);

        public void Xor(GraphicsPath path) => Combine(path, CombineMode.Xor);

        public void Xor(Region region) => Combine(region, CombineMode.Xor);

        public void Exclude(Rectangle rect) => Combine(rect, CombineMode.Exclude);

        public void Exclude(RectangleF rect) => Combine(rect, CombineMode.Exclude);

        public void Exclude(GraphicsPath path) => Combine(path, CombineMode.Exclude);

        public void Exclude(Region region) => Combine(region, CombineMode.Exclude);

        public void Complement(Rectangle rect) => Combine(rect, CombineMode.Complement);

        public void Complement(RectangleF rect) => Combine(rect, CombineMode.Complement);

        public void Complement(GraphicsPath path) => Combine(path, CombineMode.Complement);

        public void Complement(Region region) => Combine(region, CombineMode.Complement);

        public void Translate(float dx, float dy) => Transform(new Matrix(1, 0, 0, 1, dx, dy));

        public void Translate(int dx, int dy) => Translate((float)dx, dy);

        public void Transform(Matrix matrix)
        {
            if (matrix == null)
            {
                throw new ArgumentNullException("matrix");
            }
            for (int i = 0; i < steps.Count; i++)
            {
                PointF[] points = steps[i].Points;
                if (points == null)
                {
                    continue;
                }
                for (int j = 0; j < points.Length; j++)
                {
                    points[j] = matrix.Apply(points[j].X, points[j].Y);
                }
            }
        }

        // Бесконечная область у GDI+ — только та, что ни с чем не сложена.
        public bool IsInfinite(Graphics g) => steps.Count == 1 && steps[0].Kind == KindInfinite;

        public bool IsEmpty(Graphics g) => Bounds(out RectangleF bounds) == 1;

        public bool IsVisible(float x, float y) => GdiNative.RegionContains(Program(out float[] points, out byte[] types), points, types, x, y);

        public bool IsVisible(float x, float y, Graphics g) => IsVisible(x, y);

        public bool IsVisible(PointF point) => IsVisible(point.X, point.Y);

        public bool IsVisible(PointF point, Graphics g) => IsVisible(point.X, point.Y);

        public bool IsVisible(int x, int y, Graphics g) => IsVisible((float)x, y);

        public bool IsVisible(Point point) => IsVisible((float)point.X, point.Y);

        public bool IsVisible(Point point, Graphics g) => IsVisible((float)point.X, point.Y);

        // Прямоугольник виден, если видна хоть одна его угловая или средняя
        // точка — приближение: область, пересекающая прямоугольник только
        // узкой полосой между ними, сочтётся невидимой.
        public bool IsVisible(RectangleF rect)
        {
            if (rect.Width <= 0 || rect.Height <= 0)
            {
                return false;
            }
            RectangleF bounds;
            int kind = Bounds(out bounds);
            if (kind == 1 || (kind == 2 && !bounds.IntersectsWith(rect)))
            {
                return false;
            }
            for (int i = 0; i <= 4; i++)
            {
                for (int j = 0; j <= 4; j++)
                {
                    float x = rect.X + Math.Min(rect.Width * i / 4, rect.Width - 0.01f);
                    float y = rect.Y + Math.Min(rect.Height * j / 4, rect.Height - 0.01f);
                    if (IsVisible(x, y))
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        public bool IsVisible(Rectangle rect) => IsVisible((RectangleF)rect);

        public bool IsVisible(RectangleF rect, Graphics g) => IsVisible(rect);

        public bool IsVisible(Rectangle rect, Graphics g) => IsVisible((RectangleF)rect);

        public RectangleF GetBounds(Graphics g)
        {
            if (g == null)
            {
                throw new ArgumentNullException("g");
            }
            RectangleF bounds;
            switch (Bounds(out bounds))
            {
                case 0:
                    return new RectangleF(-4194304, -4194304, 8388608, 8388608);
                case 1:
                    return RectangleF.Empty;
                default:
                    return bounds;
            }
        }

        // 0 — бесконечна, 1 — пуста, 2 — границы в `bounds`.
        private int Bounds(out RectangleF bounds)
        {
            var box = new float[4];
            int kind = GdiNative.RegionBounds(Program(out float[] points, out byte[] types), points, types, box);
            bounds = new RectangleF(box[0], box[1], box[2], box[3]);
            return kind;
        }

        // Программа для растеризатора: по пять чисел на шаг и общие точки.
        internal int[] Program(out float[] points, out byte[] types)
        {
            int total = 0;
            for (int i = 0; i < steps.Count; i++)
            {
                if (steps[i].Points != null)
                {
                    total += steps[i].Points.Length;
                }
            }
            points = new float[total * 2];
            types = new byte[total];
            var ops = new int[steps.Count * 5];
            int at = 0;
            for (int i = 0; i < steps.Count; i++)
            {
                Step step = steps[i];
                int count = step.Points == null ? 0 : step.Points.Length;
                ops[i * 5] = step.Kind;
                ops[i * 5 + 1] = step.Combine;
                ops[i * 5 + 2] = at;
                ops[i * 5 + 3] = count;
                ops[i * 5 + 4] = step.FillMode;
                for (int j = 0; j < count; j++)
                {
                    points[(at + j) * 2] = step.Points[j].X;
                    points[(at + j) * 2 + 1] = step.Points[j].Y;
                    types[at + j] = step.Types[j];
                }
                at += count;
            }
            return ops;
        }

        // Прямоугольник ли область — тогда отсечение обходится без маски.
        // Узнаются бесконечность, прямоугольник со сторонами вдоль осей и
        // пересечения таких; всё остальное — «не прямоугольник», даже если
        // геометрически он (объединение двух половин, например).
        internal bool TryGetRectangle(out RectangleF rect, out bool infinite)
        {
            var stack = new List<RectangleF>();
            var flags = new List<bool>();
            rect = RectangleF.Empty;
            infinite = false;
            for (int i = 0; i < steps.Count; i++)
            {
                Step step = steps[i];
                if (step.Kind == KindInfinite)
                {
                    stack.Add(RectangleF.Empty);
                    flags.Add(true);
                }
                else if (step.Kind == KindEmpty)
                {
                    stack.Add(RectangleF.Empty);
                    flags.Add(false);
                }
                else if (step.Kind == KindPath)
                {
                    if (!IsAxisRectangle(step.Points, step.Types, out RectangleF r))
                    {
                        return false;
                    }
                    stack.Add(r);
                    flags.Add(false);
                }
                else
                {
                    if (step.Combine != (int)CombineMode.Intersect || stack.Count < 2)
                    {
                        return false;
                    }
                    RectangleF b = stack[stack.Count - 1];
                    bool bInfinite = flags[flags.Count - 1];
                    RectangleF a = stack[stack.Count - 2];
                    bool aInfinite = flags[flags.Count - 2];
                    stack.RemoveAt(stack.Count - 1);
                    flags.RemoveAt(flags.Count - 1);
                    RectangleF result = aInfinite ? b : bInfinite ? a : RectangleF.Intersect(a, b);
                    stack[stack.Count - 1] = result;
                    flags[flags.Count - 1] = aInfinite && bInfinite;
                }
            }
            if (stack.Count != 1)
            {
                return false;
            }
            rect = stack[0];
            infinite = flags[0];
            return true;
        }

        private static bool IsAxisRectangle(PointF[] p, byte[] types, out RectangleF rect)
        {
            rect = RectangleF.Empty;
            if (p == null || p.Length != 4 || types[1] != 1 || types[2] != 1 || (types[3] & 7) != 1)
            {
                return false;
            }
            bool horizontalFirst = p[0].Y == p[1].Y && p[1].X == p[2].X && p[2].Y == p[3].Y && p[3].X == p[0].X;
            bool verticalFirst = p[0].X == p[1].X && p[1].Y == p[2].Y && p[2].X == p[3].X && p[3].Y == p[0].Y;
            if (!horizontalFirst && !verticalFirst)
            {
                return false;
            }
            float x0 = Math.Min(p[0].X, p[2].X);
            float y0 = Math.Min(p[0].Y, p[2].Y);
            rect = new RectangleF(x0, y0, Math.Max(p[0].X, p[2].X) - x0, Math.Max(p[0].Y, p[2].Y) - y0);
            return true;
        }
    }
}

File: crates/raster/src/region.rs
//! Области отсечения: `Region` и `Graphics.Clip` как программа операций.
//!
//! `Region` в GDI+ — не прямоугольник, а результат операций над путями:
//! объединить с эллипсом, вычесть прямоугольник, исключающее «или» с другой
//! областью. C# хранит эти операции как есть (`System.Drawing.Region`) в
//! обратной польской записи: фигура кладётся на стек, операция снимает две
//! верхние и кладёт результат. Запись, а не цепочка «слева направо», потому
//! что `a.Union(b)`, где `b` сама собрана из трёх фигур, цепочкой не
//! выражается. Здесь программа превращается в маску точек ровно там, где идёт
//! рисование, — или проверяется в одной точке для `IsVisible`.
//!
//! Маска без сглаживания: у GDI+ область отсекает по точкам целиком, и край
//! эллипса, которым отсечена заливка, выходит ступенчатым даже при
//! `SmoothingMode.AntiAlias`.

use alloc::vec::Vec;

use crate::geom::{Matrix, Point, flatten};
use crate::scan::{self, Bounds, FillRule};
use crate::{Error, filled, push};

/// Покрытие точек прямоугольника: 0 — отсечено, 255 — видно.
pub struct Mask {
    bounds: Bounds,
    data: Vec<u8>,
}

impl Mask {
    #[must_use]
    pub const fn bounds(&self) -> Bounds {
        self.bounds
    }

    /// Покрытие точки; снаружи прямоугольника — ноль.
    #[must_use]
    pub fn at(&self, x: i32, y: i32) -> u8 {
        if x < self.bounds.x0 || y < self.bounds.y0 || x >= self.bounds.x1 || y >= self.bounds.y1 {
            return 0;
        }
        let index = (y - self.bounds.y0) as usize * self.bounds.width() + (x - self.bounds.x0) as usize;
        self.data.get(index).copied().unwrap_or(0)
    }
}

/// `CombineMode`.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Combine {
    Replace,
    Intersect,
    Union,
    Xor,
    Exclude,
    Complement,
}

impl Combine {
    /// Код `System.Drawing.Drawing2D.CombineMode`.
    #[must_use]
    pub const fn from_code(code: i32) -> Self {
        match code {
            1 => Self::Intersect,
            2 => Self::Union,
            3 => Self::Xor,
            4 => Self::Exclude,
            5 => Self::Complement,
            _ => Self::Replace,
        }
    }

    const fn apply(self, a: bool, b: bool) -> bool {
        match self {
            Self::Replace => b,
            Self::Intersect => a && b,
            Self::Union => a || b,
            Self::Xor => a != b,
            Self::Exclude => a && !b,
            Self::Complement => b && !a,
        }
    }
}

/// Шаг программы.
#[derive(Clone, Copy)]
pub enum Element<'a> {
    Infinite,
    Empty,
    Path { points: &'a [f32], types: &'a [u8], rule: FillRule },
    Combine(Combine),
}

/// Прочитать программу, как её раскладывает C#: по пять чисел на шаг —
/// `вид (0 бесконечность, 1 пусто, 2 путь, 3 операция), CombineMode, первая
/// точка, число точек, FillMode` — и общие для всех путей точки и виды точек.
/// Путь с диапазоном за пределами массивов считается пустым: это данные
/// программы, и падать на них среда не вправе.
pub fn decode<'a>(ops: &[i32], points: &'a [f32], types: &'a [u8]) -> Result<Vec<Element<'a>>, Error> {
    let mut elements = Vec::new();
    for op in ops.chunks_exact(5) {
        let element = match op[0] {
            0 => Element::Infinite,
            1 => Element::Empty,
            3 => Element::Combine(Combine::from_code(op[1])),
            _ => {
                let range = usize::try_from(op[2]).ok().zip(usize::try_from(op[3]).ok()).and_then(|(start, count)| {
                    let end = start.checked_add(count)?;
                    (end <= types.len() && end.checked_mul(2)? <= points.len()).then_some((start, end))
                });
                match range {
                    Some((start, end)) => Element::Path {
                        points: &points[start * 2..end * 2],
                        types: &types[start..end],
                        rule: if op[4] == 1 { FillRule::NonZero } else { FillRule::EvenOdd },
                    },
                    None => Element::Empty,
                }
            }
        };
        push(&mut elements, element)?;
    }
    Ok(elements)
}

fn polygons(points: &[f32], types: &[u8], matrix: &Matrix) -> Result<Vec<Vec<Point>>, Error> {
    let mut out = Vec::new();
    for figure in flatten(points, types, matrix, 0.25)? {
        push(&mut out, figure.points)?;
    }
    Ok(out)
}

/// Маска программы в прямоугольнике `bounds` после преобразования `matrix`
/// (координаты области → пространство растеризатора). Пустая программа —
/// бесконечная область; лишнее на стеке после конца объединяется, как если
/// бы программа кончалась объединением (так её не построит C#, но и падать
/// тут не на чем).
pub fn mask(elements: &[Element<'_>], matrix: &Matrix, bounds: Bounds) -> Result<Mask, Error> {
    let size = bounds.width() * bounds.height();
    let width = bounds.width();
    let mut stack: Vec<Vec<u8>> = Vec::new();
    for element in elements {
        match *element {
            Element::Infinite => push(&mut stack, filled(size, 255u8)?)?,
            Element::Empty => push(&mut stack, filled(size, 0u8)?)?,
            Element::Path { points, types, rule } => {
                let mut shape = filled(size, 0u8)?;
                let polygons = polygons(points, types, matrix)?;
                scan::rasterize(&polygons, rule, false, bounds, &mut |y, x, row| {
                    let base = (y - bounds.y0) as usize * width + (x - bounds.x0) as usize;
                    if let Some(slots) = shape.get_mut(base..base + row.len()) {
                        slots.copy_from_slice(row);
                    }
                })?;
                push(&mut stack, shape)?;
            }
            Element::Combine(combine) => {
                let Some(b) = stack.pop() else { continue };
                let Some(mut a) = stack.pop() else {
                    push(&mut stack, b)?;
                    continue;
                };
                for (a, &b) in a.iter_mut().zip(b.iter()) {
                    *a = if combine.apply(*a != 0, b != 0) { 255 } else { 0 };
                }
                push(&mut stack, a)?;
            }
        }
    }
    let mut data = match stack.pop() {
        Some(data) => data,
        None => filled(size, 255u8)?,
    };
    while let Some(rest) = stack.pop() {
        for (a, &b) in data.iter_mut().zip(rest.iter()) {
            *a = (*a).max(b);
        }
    }
    Ok(Mask { bounds, data })
}

/// Лежит ли точка области в программе — пересечениями, без растеризации.
pub fn contains(elements: &[Element<'_>], x: f64, y: f64) -> Result<bool, Error> {
    let mut stack: Vec<bool> = Vec::new();
    for element in elements {
        match *element {
            Element::Infinite => push(&mut stack, true)?,
            Element::Empty => push(&mut stack, false)?,
            Element::Path { points, types, rule } => {
                let polygons = polygons(points, types, &Matrix::IDENTITY)?;
                push(&mut stack, scan::contains(&polygons, rule, x, y))?;
            }
            Element::Combine(combine) => {
                let b = stack.pop().unwrap_or(false);
                let a = stack.pop().unwrap_or(false);
                push(&mut stack, combine.apply(a, b))?;
            }
        }
    }
    Ok(stack.iter().any(|inside| *inside) || elements.is_empty())
}

/// Границы области.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Extent {
    Infinite,
    Empty,
    Box([f64; 4]),
}

/// Границы программы оценкой по прямоугольникам: объединение и «или» —
/// объединение границ, пересечение — пересечение, вычитание — границы
/// уменьшаемого. Совпадает с `Region.GetBounds` у GDI+ на пробе (объединение,
/// потом пересечение, потом «или» и вычитание угла); точная граница после
/// вычитания, отрезающего целую сторону, потребовала бы растеризации.
pub fn extent(elements: &[Element<'_>]) -> Result<Extent, Error> {
    let mut stack: Vec<Extent> = Vec::new();
    for element in elements {
        let value = match *element {
            Element::Infinite => Extent::Infinite,
            Element::Empty => Extent::Empty,
            Element::Path { points, .. } => {
                let mut b: Option<[f64; 4]> = None;
                for pair in points.chunks_exact(2) {
                    let (x, y) = (f64::from(pair[0]), f64::from(pair[1]));
                    if !x.is_finite() || !y.is_finite() {
                        continue;
                    }
                    b = Some(match b {
                        None => [x, y, x, y],
                        Some([x0, y0, x1, y1]) => [x0.min(x), y0.min(y), x1.max(x), y1.max(y)],
                    });
                }
                match b {
                    Some(b) if b[2] > b[0] && b[3] > b[1] => Extent::Box(b),
                    _ => Extent::Empty,
                }
            }
            Element::Combine(combine) => {
                let b = stack.pop().unwrap_or(Extent::Empty);
                let a = stack.pop().unwrap_or(Extent::Empty);
                combine_extent(combine, a, b)
            }
        };
        push(&mut stack, value)?;
    }
    let mut result = stack.pop().unwrap_or(Extent::Infinite);
    while let Some(rest) = stack.pop() {
        result = combine_extent(Combine::Union, rest, result);
    }
    Ok(result)
}

fn combine_extent(combine: Combine, a: Extent, b: Extent) -> Extent {
    match combine {
        Combine::Replace => b,
        Combine::Intersect => match (a, b) {
            (Extent::Infinite, other) | (other, Extent::Infinite) => other,
            (Extent::Box(a), Extent::Box(b)) => {
                let r = [a[0].max(b[0]), a[1].max(b[1]), a[2].min(b[2]), a[3].min(b[3])];
                if r[2] > r[0] && r[3] > r[1] { Extent::Box(r) } else { Extent::Empty }
            }
            _ => Extent::Empty,
        },
        Combine::Union | Combine::Xor => match (a, b) {
            (Extent::Infinite, _) | (_, Extent::Infinite) => Extent::Infinite,
            (Extent::Empty, other) | (other, Extent::Empty) => other,
            (Extent::Box(a), Extent::Box(b)) => Extent::Box([a[0].min(b[0]), a[1].min(b[1]), a[2].max(b[2]), a[3].max(b[3])]),
        },
        Combine::Exclude => a,
        Combine::Complement => b,
    }
}
